### PR TITLE
Disabled the header checkbox to (un)select all lines in DataTable

### PR DESCRIPTION
**Pull request by LegerG**
#### What does this PR do?

Add an option to disable the "Select All" checkbox in the header when `select` or `onSelect` props are set in `DataTable`
This prop is named `allSelectedDisabled`. I'm not sure about the naming, don't hesitate to share yours ;)

#### Where should the reviewer start?

src/js/components/DataTable/DataTable.js

#### What testing has been done on this PR?

Only manual testing with storybook

#### Do Jest tests follow these best practices?

- [ ] `screen` is used for querying.
- [ ] The correct query is used. (Refer to [this list of queries](https://testing-library.com/docs/queries/about/#priority))
- [ ] `userEvent` is used in place of `fireEvent`.
- [ ] `asFragment()` is used for snapshot testing.

#### Any background context you want to provide?

I have this need in my projects, and see there is an open issue.

#### What are the relevant issues?

https://github.com/grommet/grommet/issues/6843

#### Screenshots (if appropriate)

image

#### Do the grommet docs need to be updated?

Yes

#### Should this PR be mentioned in the release notes?

Yes

#### Is this change backwards compatible or is it a breaking change?

Backward compatible

### CODE REVIEW

**Comment by taysea** [comment]
Thanks for filing this PR! We are having a few discussions on the design direction/experience of this kind of flow. Once we have more clear direction there, we'll drop any comments here for items to be adjusted.

**Comment by LegerG** [comment]
Thanks for your response, I keep in touch.

**Comment by jcfilben** [comment]
Hey @LegerG the team evaluated the use case in the original ticket (#6843) and realized that for that particular case where we limit the number of rows that can be selected we wanted to go a different direction than removing the select all checkbox. However, I saw in your PR description you mentioned that you needed this functionality for one of your projects. Would you be willing to share your use case details? The reason I ask is we like any enhancements that get added to grommet to be use case driven. This helps ensure we keep the grommet api clean.

**Comment by LegerG** [comment]
The use case is the following:
I have a list of room reservations, each row as the id of the reservation, the id of the room, the datetime and the duration.
I want to select some of them with a limited criterion based on the room id. If I select the room number 12 at 4pm, I want my user to only select a reservation for the room number 12, and not for others.
In this case, the "all select" checkbox is not wanted because some rows must not be selected together.
I think this use case could be generalized as a qualitative restriction of selection, and not quantitative as the limitation on the number of rows.

For now, I have used CSS tricks to remove the checkbox and manage which lines could be selected or not. 

I hope this explanation has enough details.

**Comment by jcfilben** [comment]
Thanks for sharing your use case! I'm comfortable with moving forward with this feature. @taysea @MikeKingdom @halocline any concerns with adding this?

File: src/js/components/DataTable/DataTable.js [hunk 1]
```diff
@@ -87,6 +87,7 @@ const DataTable = ({
   rowDetails,
   step = 50,
   verticalAlign,
+  allSelectedDisabled = false,
```
Comment (jcfilben) [inline src/js/components/DataTable/DataTable.js:90]:
```suggestion
  allowSelectAll = true,
```
Can we change the prop name here to `allowSelectAll`?

**Review by jcfilben — COMMENTED** [review]
Can we add a jest test for this?
Also can we add the new prop to the `DataTable/propTypes.js` and `DataTable/index.d.ts` files?

**Comment by LegerG** [comment]
I have written a test, added types, and renamed the field :)

File: src/js/components/DataTable/index.d.ts [hunk 2]
```diff
@@ -142,6 +142,7 @@ export interface DataTableProps<TRowType = any> {
         body?: VerticalAlignType;
         footer?: VerticalAlignType;
       };
+  allowSelectAll?: boolean;
```
Comment (jcfilben) [inline src/js/components/DataTable/index.d.ts:145]: Super minor comment but can we alphabetize this prop (it should appear under `alignSelf` and above `background`)

File: src/js/components/DataTable/propTypes.js [hunk 3]
```diff
@@ -179,6 +179,7 @@ if (process.env.NODE_ENV !== 'production') {
         footer: PropTypes.oneOf(verticalAlign),
       }),
     ]),
+    allowSelectAll: PropTypes.bool,
```
Comment (jcfilben) [inline src/js/components/DataTable/propTypes.js:182]: Same request here, can we alphabetize this so it appears above `background`

File: src/js/components/DataTable/stories/AllowSelectAll.js [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+        select={select}
+        onSelect={(selected) => setSelect(selected)}
+        allowSelectAll={allowSelectAll}
+      />
+    </Box>
+  );
+};
+
+AllowSelectAll.storyName = 'Allow Select All';
+
```
Comment (jcfilben) [inline src/js/components/DataTable/stories/AllowSelectAll.js:100]:
```suggestion

AllowSelectAll.parameters = {
  chromatic: { disable: true },
};

```
Can we disable this story from chromatic snapshots?

**Review by jcfilben — COMMENTED** [review]
Just some minor comments otherwise looks good

**Review by jcfilben — APPROVED** [review]
Looks good!

**Pull request merged**